### PR TITLE
[System.Windows.Forms] Unify layout engines for Controls and ToolStripItems

### PR DESCRIPTION
**Pull request by filipnavara**
Lot of the ToolStrip layout code was held by duct tape. This PR introduces the IArrangedElement and IArrangedContainer interfaces that are implmented by Control, ToolStripItem and other ToolStrip classes. This facilitates sharing the same layout engine code for both regular controls and the tool strips. The benefit is that it enables use of table layout in ToolStrips (as used by StatusStrip) and proper use of FlowLayout in ToolStripDropDown/ToolStripDropDownMenu/ToolStripOverflow. It makes the overflow code working for KeePass on X11 backend, properly flowing the buttons and combo box from the tool strip into the overflow window.

### CODE REVIEW

**Comment by filipnavara** [comment]
The duct tape layout is still kept for ToolStripPanel and ToolStripPanelRow since I had no code to test it.

**Comment by filipnavara** [comment]
As for the API diff, there are two options. I can either keep it as-is, the public API doesn't change, only protected overridden methods, or I can move the code from CreateLayoutSettings into OnLayout. The later option is less clear in the intent.

File: mcs/class/System.Windows.Forms/System.Windows.Forms.Layout/DefaultLayout.cs [hunk 1]
```diff
@@ -43,33 +43,33 @@ private DefaultLayout ()
 	
 		public override void InitLayout (object child, BoundsSpecified specified)
 		{
-			Control control = child as Control;
-			Control parent = control.Parent;
+			IArrangedElement control = child as IArrangedElement;
```
Comment (marek-safar) [inline mcs/class/System.Windows.Forms/System.Windows.Forms.Layout/DefaultLayout.cs]: Use IArrangedElement explicit cast

File: mcs/class/System.Windows.Forms/System.Windows.Forms.Layout/FlowLayout.cs [hunk 2]
```diff
@@ -47,24 +47,26 @@ public override void InitLayout (object child, BoundsSpecified specified)
 			base.InitLayout (child, specified);
 		}
 
+		private FlowLayoutSettings GetLayoutSettings (IArrangedContainer container)
```
Comment (marek-safar) [inline mcs/class/System.Windows.Forms/System.Windows.Forms.Layout/FlowLayout.cs]: should be static

File: mcs/class/System.Windows.Forms/System.Windows.Forms.Layout/FlowLayout.cs [hunk 3]
```diff
@@ -47,24 +47,26 @@ public override void InitLayout (object child, BoundsSpecified specified)
...
+		{
+			if (container is FlowLayoutPanel)
```
Comment (marek-safar) [inline mcs/class/System.Windows.Forms/System.Windows.Forms.Layout/FlowLayout.cs]: Use `if (container is FlowLayoutPanel flp)` syntax

File: mcs/class/System.Windows.Forms/System.Windows.Forms.Layout/FlowLayout.cs [hunk 4]
```diff
@@ -47,24 +47,26 @@ public override void InitLayout (object child, BoundsSpecified specified)
...
-				return LayoutToolStrip ((ToolStrip)container);
-				
-			Control parent = container as Control;
-			
-			FlowLayoutSettings settings;
-			if (parent is FlowLayoutPanel)
-				settings = (parent as FlowLayoutPanel).LayoutSettings;
-			else
-				settings = default_settings;
+			IArrangedContainer parent = container as IArrangedContainer;
```
Comment (marek-safar) [inline mcs/class/System.Windows.Forms/System.Windows.Forms.Layout/FlowLayout.cs]: Change it to explicit cast

**Comment by marek-safar** [comment]
API diff changes are fine, they are not breaking changes

**Comment by filipnavara** [comment]
Thanks for review! I updated the code with appropriate fixes.

**Comment by marek-safar** [comment]
@monojenkins commit apidiff

**Pull request merged**